pageserver/src/virtual_file.rs

```diff
@@ -601,23 +602,36 @@ impl VirtualFile {
         Ok(())
     }
 
-    pub async fn write_all(&mut self, mut buf: &[u8]) -> Result<(), Error> {
+    /// Writes `buf.slice(0..buf.bytes_init())`.
+    /// Returns the IoBuf that is underlying the BoundedBuf `buf`.
+    /// I.e., the returned value's `bytes_init()` method returns something different than the `bytes_init()` that was passed in.
+    /// It's quite brittle and easy to mis-use, so, we return the size in the Ok() variant.
+    pub async fn write_all<B: BoundedBuf>(&mut self, buf: B) -> (B::Buf, Result<usize, Error>) {
+        let nbytes = buf.bytes_init();
+        if nbytes == 0 {
+            return (Slice::into_inner(buf.slice_full()), Ok(0));
```

Comment:
```suggestion
            return (buf, Ok(0));
```

Reply:
also wondering why this is required at all, we already have `!buf.is_empty()` check below. If it is required, then a comment would be helpful

Reply:
The pattern is to return the underlying buf, because once you start using `.slice()`, you can't go back to the previous slice, because `.slice()` takes ownership and discards the previous slice's .begin and .end.

There's no single good place to put a comment like this.

How about I can start a Notion doc with tokio-epoll-uring API conventions & commit that to `docs/` once the integration is complete?

Rationale for a Notion doc compared to updating `docs/` in each commit: there are many commits in flight right now, it's easier to update it once, in the end.